File: internal/service/category_service.go
package service

import (
	"money-manager/internal/domain"
	"money-manager/internal/dto"
	"money-manager/internal/repository"
)

type CategoryService struct {
	repo *repository.CategoryRepository
}

func NewCategoryService(repo *repository.CategoryRepository) *CategoryService {
	return &CategoryService{repo: repo}
}

func (s *CategoryService) CreateCategory(userID string, req dto.CreateCategoryRequest) (*dto.CategoryResponse, error) {
	category := &domain.Category{
		UserID:   userID,
		Name:     req.Name,
		ParentID: req.ParentID,
		IsSystem: false,
	}

	if err := s.repo.Create(category); err != nil {
		return nil, err
	}

	return &dto.CategoryResponse{
		ID:        category.ID,
		Name:      category.Name,
		ParentID:  category.ParentID,
		IsSystem:  category.IsSystem,
		CreatedAt: category.CreatedAt.Format("2006-01-02 15:04:05"),
	}, nil
}

func (s *CategoryService) GetUserCategories(userID string) ([]dto.CategoryResponse, error) {
	categories, err := s.repo.GetAllByUserID(userID)
	if err != nil {
		return nil, err
	}

	var res []dto.CategoryResponse
	for _, c := range categories {
		res = append(res, dto.CategoryResponse{
			ID:        c.ID,
			Name:      c.Name,
			ParentID:  c.ParentID,
			IsSystem:  c.IsSystem,
			CreatedAt: c.CreatedAt.Format("2006-01-02 15:04:05"),
		})
	}

	return res, nil
}

func (s *CategoryService) DeleteCategory(id string, userID string) error {
	return s.repo.Delete(id, userID)
}

File: money-manager-frontend/src/services/category.service.ts
import api from "./api";
import { ApiResponse } from "@/types";

export interface Category {
  id: string;
  name: string;
  parent_id?: string;
  is_system: boolean;
  created_at: string;
}

export const categoryService = {
  getAll: async (params?: any) => {
    const response = await api.get<ApiResponse<Category[]>>("/api/categories", { params });
    return response.data;
  },
  create: async (data: { name: string; parent_id?: string }) => {
    const response = await api.post<ApiResponse<Category>>("/api/categories", data);
    return response.data;
  },
  delete: async (id: string) => {
    const response = await api.delete<ApiResponse<null>>(`/api/categories/${id}`);
    return response.data;
  },
};
